oeps/oep-0026-arch-xapi-integration.rst

```diff
@@ -0,0 +1,421 @@
+.. _Experience API (xAPI): https://www.adlnet.gov/research/performance-tracking-analysis/experience-api/
+.. _widespread industry support and usage: https://xapi.com/adopters/
+.. _xAPI Architecture Overview: https://www.adlnet.gov/research/performance-tracking-analysis/experience-api/xapi-architecture-overview/
+
+Learning Record Store (LRS) - Future
+====================================
+
+xAPI includes specification for a Learning Record Store (LRS), which encapsulates a
+data store with APIs for storing and accessing xAPI data. While an LRS could be used
+for many different purposes, it can be used to train adaptive engines as shown in the
+diagram below.
+
+.. image:: oep-0026/adaptive_learning_lrs_basic.png
+   :alt: The diagram above is enhanced with a new LRS component that is sent events
+    (write requests) and accessed by adaptive engines (read requests) for training
+    purposes, using xAPI for both read and write.
+
+In the short term, however, we will not implement our own LRS since the initial integration
+efforts do not require this (as shown in the first diagram above).
+
+Statement API
+=============
+
+xAPI specifies 4 standard REST-ful JSON payload APIs, of which only the `Statement API`_
+is needed for our purposes as that API is sufficient for tracking learning activity.
+
+The Statement API has essentially the following parts: `xAPI Actor`_, `xAPI Verb`_,
+`xAPI Object`_, `xAPI Context`_, and `xAPI Result`_.
+
+.. _Statement API: https://xapi.com/statements-101/
+
+xAPI Actor
+~~~~~~~~~~
+Although the **Actor** field can be either an *Agent* or a *Group*, we will primarily
+support only the *Agent* type, which is used for individuals performing an action (**Verb** 
+on **Object**).
+
+An Actor can be identified using a `Friend of a Friend (FOAF)`_ vocabulary with either:
+(1) `email address`_, (2) `hash of email address`_, (3) `OpenID URI`_, or (4) `account`_
+with a *homepage*-scoped identifier.  One of these is sent along with the Actor's "name".
+To be mindful of learner privacy, we will initially take a conservative approach
+and only send #3, with an Open edX randomly generated unique identifier of the learner
+(`Open edX User UUID`_).
+
+In the future, if certain external systems require `Personally Identifiable Information
+(PII)`_, like the learner's email address or name, then those may be conditionally sent
+with appropriate permissions. Adaptive engines, however, do not need PII.
+
+Initially, we will exclude the "name" field. However, if we find that xAPI JSON parsers
+assume this field always exists, then we can include the field but provide a non-PII
+value, such as a copy of the `Open edX User UUID`_.
+
+Open edX User UUID
+^^^^^^^^^^^^^^^^^^
+An *Open edX User UUID* identifier currently doesn't exist in the platform. It will need 
+to be added as part of implementing xAPI support. There would be a 1:1 mapping between
+a randomly generated UUID_ and a row in the Django User table in the Open edX LMS.
```

Comment:
I have performance-related concerns with this 1:1 relationship. Using a custom user model may not be also considered?

Reply:
We will need to worry about backward compatibility with large-scale Open edX deployments (like at edx.org) where a migration of a very large auth_user table is not feasible with our current MySQL technology offering.

Another short-term alternative - I've been thinking about - in order to quickly move forward in the first iteration - instead of investing time in this new UUID join table, we could use the `hash value of the LMS user_id` as the unique identifier.  This value would not leak information, but it will still tie us to the database structure.  However, if we ever did migrate to a new auth_user table, we could theoretically add the new UUID column at that time and populate it with the hash of the original id values.  This would allow us to dynamically compute this value without needing to make any database migrations.  What do you think?

@edx/arch-review Any thoughts on this?

Reply:
Looks like a solution for outputting messages with user uuids but could be a limitation if you need to come back and query the user table with these IDs? Do you think this could be required or can we rule it out?

Reply:
It's a requirement to find an appropriate user identifier for xAPI events.
I didn't understand your first question.  If you need to do a reverse lookup from User UUID to User, you can do so via a SQL JOIN operation, right?

Reply:
My remark was regarding your alternative:  "we could use the hash value of the LMS user_id as the unique identifier."

I agree with @jmaupetit that adding a separate table for the UUID is not a viable solution. I'm not expert in mysql but I'm surprised that the size of a table prevents adding a new column?

Reply:
For our current edx.org deployment, migrating large SQL tables result in server downtime, which is not acceptable for our situation.

I will go ahead and update the docs to say that we'll send a hash of the LMS user_id - instead of creating a new User UUID value.